sdks/java/io/distributedlog/src/main/java/org/apache/beam/sdk/io/distributedlog/MultipleSegmentsReader.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.distributedlog;
+
+import com.google.common.collect.ImmutableList;
+import com.twitter.distributedlog.exceptions.DLInterruptedException;
+import com.twitter.distributedlog.namespace.DistributedLogNamespace;
+import java.io.IOException;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.NoSuchElementException;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.io.BoundedSource;
+import org.apache.beam.sdk.io.BoundedSource.BoundedReader;
+
+/**
+ * A bounded reader reading data from a distributedlog segment.
+ */
+class MultipleSegmentsReader<R> extends BoundedReader<R>{
+
+  private final DLBoundedSource<R> source;
+
+  private List<LogSegmentBundle> segments;
+  private ListIterator<LogSegmentBundle> segmentsIterator;
+  // read state
+  private DistributedLogNamespace namespace;
+  private SingleLogSegmentReader<R> currentReader;
+  private R currentR;
+  private volatile boolean done = false;
+
+  MultipleSegmentsReader(
+      DLBoundedSource<R> source,
+      @Nullable LogSegmentBundle bundle) {
+    this.source = source;
+    if (null != bundle) {
+      this.segments = ImmutableList.of(bundle);
+      this.segmentsIterator = segments.listIterator();
+    }
+  }
+
+  @Override
+  public boolean start() throws IOException {
```

Comment:
Make sure to extensively unit-test the Source class and the current class with SourceTestUtils, especially including corner cases with empty segments and such.